File: legacy/typescript-prototype/src/metadata.ts
// metadata 模块负责通过 well-known 文档发现 Agent 身份并做基础校验。
import { assertDomainMatch, parseAgentId } from "./identity.js";
import type { AgentKey, AgentMetadata, ResolveResult, ResolverOptions } from "./types.js";

interface CacheEntry {
  result: ResolveResult;
  expiresAt: number;
}

const cache = new Map<string, CacheEntry>();
const DEFAULT_CACHE_TTL_SECONDS = 300;

function isLocalhostDomain(domain: string): boolean {
  return domain.startsWith("localhost:") || domain.startsWith("127.0.0.1:");
}

function isIsoDate(input: string | undefined): boolean {
  if (!input) {
    return false;
  }

  return !Number.isNaN(Date.parse(input));
}

function validateMetadata(metadata: AgentMetadata): void {
  // 这里做的是运行时校验，目标是即使没有额外 schema 库也能尽早发现脏数据。
  if (!metadata || typeof metadata !== "object") {
    throw new Error("metadata must be an object");
  }

  if (!metadata.agent_id || !metadata.domain || !metadata.endpoint) {
    throw new Error("metadata is missing required fields");
  }

  assertDomainMatch(metadata.agent_id, metadata.domain);

  const endpoint = new URL(metadata.endpoint);
  const isLocalhost = endpoint.hostname === "localhost" || endpoint.hostname === "127.0.0.1";
  // 正式环境强制 HTTPS，本地 demo 为了方便联调允许 localhost 走 HTTP。
  if (endpoint.protocol !== "https:" && !(isLocalhost && endpoint.protocol === "http:")) {
    throw new Error("endpoint must use https, localhost demo may use http");
  }

  if (!Array.isArray(metadata.capabilities) || !Array.isArray(metadata.keys)) {
    throw new Error("metadata capabilities and keys must be arrays");
  }

  if (!Array.isArray(metadata.revoked_kids)) {
    throw new Error("metadata revoked_kids must be an array");
  }

  if (!isIsoDate(metadata.updated_at)) {
    throw new Error("metadata updated_at must be an ISO date");
  }

  const seenKids = new Set<string>();
  for (const key of metadata.keys) {
    // kid 唯一性是轮换逻辑成立的前提。
    if (!key.kid || !key.public_key || key.alg !== "Ed25519") {
      throw new Error("metadata key is invalid");
    }

    if (seenKids.has(key.kid)) {
      throw new Error(`duplicate kid: ${key.kid}`);
    }
    seenKids.add(key.kid);

    if (key.not_before && !isIsoDate(key.not_before)) {
      throw new Error(`invalid not_before: ${key.kid}`);
    }

    if (key.not_after && !isIsoDate(key.not_after)) {
      throw new Error(`invalid not_after: ${key.kid}`);
    }
  }
}

export async function resolveAgent(
  agentId: string,
  options: ResolverOptions = {},
): Promise<ResolveResult> {
  const parsed = parseAgentId(agentId);
  const fetchImpl = options.fetch ?? globalThis.fetch;
  if (!fetchImpl) {
    throw new Error("fetch implementation is required");
  }

  const ttlMs = (options.cacheTtlSeconds ?? DEFAULT_CACHE_TTL_SECONDS) * 1000;
  const cacheKey = parsed.raw;
  const cached = cache.get(cacheKey);
  const now = Date.now();

  if (cached && cached.expiresAt > now) {
    // 命中有效缓存时直接返回，避免每次验签都访问远程域名。
    return cached.result;
  }

  // localhost 联调用 http，其他场景默认 https。
  const scheme = isLocalhostDomain(parsed.domain) ? "http" : "https";
  const wellKnownUrl = `${scheme}://${parsed.domain}/.well-known/agent.json`;
  const headers: HeadersInit = {};
  if (cached?.result.etag) {
    // 如果上次有 ETag，这里尝试条件请求以减少重复拉取。
    headers["If-None-Match"] = cached.result.etag;
  }

  try {
    const response = await fetchImpl(wellKnownUrl, { headers });
    if (response.status === 304 && cached) {
      // 远端未变化时只刷新本地 TTL，不重复解析 JSON。
      const updated = {
        ...cached,
        expiresAt: now + ttlMs,
      };
      cache.set(cacheKey, updated);
      return updated.result;
    }

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const metadata = (await response.json()) as AgentMetadata;
    validateMetadata(metadata);
    assertDomainMatch(agentId, metadata.domain);

    const etag = response.headers.get("etag");
    const result: ResolveResult = etag
      ? {
          metadata,
          resolvedAt: new Date().toISOString(),
          etag,
        }
      : {
          metadata,
          resolvedAt: new Date().toISOString(),
        };

    cache.set(cacheKey, { result, expiresAt: now + ttlMs });
    return result;
  } catch (error) {
    if (cached && cached.expiresAt > now) {
      // 远端临时失败时允许回退到仍然有效的旧缓存，提升可用性。
      return cached.result;
    }

    throw error;
  }
}

export function getActiveKey(metadata: AgentMetadata, kid?: string, now = new Date()): AgentKey {
  // 如果请求明确带了 kid，就优先精确匹配，不做模糊兜底。
  if (kid && metadata.revoked_kids.includes(kid)) {
    throw new Error(`Key revoked: ${kid}`);
  }

  const matches = metadata.keys.filter((key) => {
    if (kid && key.kid !== kid) {
      return false;
    }

    if (metadata.revoked_kids.includes(key.kid)) {
      return false;
    }

    if (key.status !== "active") {
      return false;
    }

    // active + 时间窗 + 未吊销，共同决定这把 key 是否可用于当前验签。
    const current = now.getTime();
    const notBefore = key.not_before ? Date.parse(key.not_before) : undefined;
    const notAfter = key.not_after ? Date.parse(key.not_after) : undefined;

    if (notBefore !== undefined && notBefore > current) {
      return false;
    }

    if (notAfter !== undefined && notAfter < current) {
      return false;
    }

    return true;
  });

  if (matches.length === 0) {
    throw new Error(kid ? `Active key not found for kid=${kid}` : "No active key found");
  }

  return matches[0]!;
}

export function clearMetadataCache(): void {
  // 测试里会主动清缓存，避免不同 case 之间互相污染。
  cache.clear();
}

File: agent_auth_sdk/metadata.py
"""metadata 发现、校验与公钥选择逻辑。"""

from __future__ import annotations

import asyncio
import ipaddress
import socket
from datetime import UTC, datetime
from urllib.parse import urlencode, urlparse, urlunparse

import httpx

from .config import STRICT_PROFILE, DiscoveryMode, MetadataResolverConfig, RuntimeProfile
from .errors import AgentIdentityError, MetadataValidationError
from .identity import assert_subject_match, build_agent_id, parse_agent_id
from .models import AgentKey, AgentMetadata, AgentRegistryDocument, ResolveResult
from .stores import MetadataCache


def _split_host_port(host: str) -> tuple[str, str | None]:
    parsed = urlparse(f"//{host}")
    try:
        port = str(parsed.port) if parsed.port is not None else None
    except ValueError:
        return host, None
    return parsed.hostname or host, port


def _is_ip_host(host: str) -> bool:
    hostname, _ = _split_host_port(host)
    try:
        ipaddress.ip_address(hostname)
        return True
    except ValueError:
        return hostname.lower() == "localhost"


def metadata_url_for_agent(agent_id: str, profile: RuntimeProfile) -> str:
    parsed = parse_agent_id(agent_id)
    scheme = "http" if profile.allow_http and _is_ip_host(parsed.host) else "https"
    return f"{scheme}://{parsed.host}/.well-known/agent.json"


def validate_metadata(metadata: AgentMetadata, profile: RuntimeProfile) -> None:
    assert_subject_match(metadata.agent_id, metadata.domain)
    if build_agent_id(metadata.domain, metadata.name) != metadata.agent_id:
        raise MetadataValidationError("metadata domain/name do not match agent_id")
    endpoint = urlparse(metadata.endpoint)
    if not endpoint.scheme or not endpoint.netloc:
        raise MetadataValidationError("endpoint must be an absolute URL")
    endpoint_is_ip = _is_ip_host(endpoint.netloc)
    if endpoint.scheme != "https":
        if not (profile.allow_http and endpoint.scheme == "http" and endpoint_is_ip):
            raise MetadataValidationError("endpoint must use https in current profile")
    if not profile.allow_ip_host and _is_ip_host(metadata.domain):
        raise MetadataValidationError("IP/localhost host is rejected in strict profile")

    seen_kids: set[str] = set()
    for key in metadata.keys:
        if key.kid in seen_kids:
            raise MetadataValidationError(f"duplicate kid: {key.kid}")
        seen_kids.add(key.kid)


async def resolve_agent(
    agent_id: str,
    *,
    profile: RuntimeProfile | None = None,
    http_client: httpx.AsyncClient,
    cache: MetadataCache | None = None,
    config: MetadataResolverConfig | None = None,
) -> ResolveResult:
    config = config or MetadataResolverConfig(profile=profile or STRICT_PROFILE)
    profile = profile or config.profile
    ttl_seconds = config.cache_ttl_seconds or profile.metadata_cache_ttl_seconds
    parse_agent_id(agent_id)

    cached = await cache.get(agent_id) if cache else None
    if cached:
        return cached

    mode = config.effective_discovery_mode
    if mode in {DiscoveryMode.REGISTRY_ONLY, DiscoveryMode.REGISTRY_THEN_DIRECT}:
        if not config.registry_url:
            raise MetadataValidationError("registry_url is required for registry discovery")
        try:
            result = await _resolve_from_registry(
                agent_id,
                registry_url=config.registry_url,
                profile=profile,
                http_client=http_client,
                timeout_seconds=config.request_timeout_seconds,
            )
            if cache:
                await cache.set(agent_id, result, ttl_seconds)
            return result
        except Exception:
            if mode == DiscoveryMode.REGISTRY_ONLY:
                raise

    if mode not in {DiscoveryMode.DIRECT_ONLY, DiscoveryMode.REGISTRY_THEN_DIRECT}:
        raise MetadataValidationError(f"unsupported discovery mode: {mode}")
    url = metadata_url_for_agent(agent_id, profile)
    pinned_ip = await _validate_direct_target(url, profile)
    request_url = _pin_direct_url(url, pinned_ip)
    parsed_direct_url = urlparse(url)
    request_headers = {"host": parsed_direct_url.netloc}
    request_extensions = (
        {"sni_hostname": parsed_direct_url.hostname}
        if parsed_direct_url.scheme == "https" and pinned_ip != parsed_direct_url.hostname
        else None
    )
    try:
        response = await http_client.get(
            request_url,
            headers=request_headers,
            timeout=config.request_timeout_seconds,
            follow_redirects=False,
            extensions=request_extensions,
        )
        response.raise_for_status()
        payload = response.json()
        if isinstance(payload, dict) and payload.get("registry_type") == "agent_registry":
            document = AgentRegistryDocument.model_validate(payload)
            entry = next((item for item in document.agents if item.agent_id == agent_id), None)
            if entry is None:
                raise MetadataValidationError(f"agent not found in well-known document: {agent_id}")
            metadata = entry.metadata
        else:
            metadata = AgentMetadata.model_validate(payload)
        if metadata.agent_id != agent_id:
            raise MetadataValidationError("resolved metadata agent_id does not match requested agent_id")
        validate_metadata(metadata, profile)
        assert_subject_match(agent_id, metadata.domain)
        result = ResolveResult(
            metadata=metadata,
            resolved_at=datetime.now(UTC),
            etag=response.headers.get("etag"),
            source_url=url,
        )
        if cache:
            await cache.set(agent_id, result, ttl_seconds)
        return result
    except (httpx.HTTPError, ValueError, AgentIdentityError, MetadataValidationError) as exc:
        raise exc


async def _resolve_from_registry(
    agent_id: str,
    *,
    registry_url: str,
    profile: RuntimeProfile,
    http_client: httpx.AsyncClient,
    timeout_seconds: float,
) -> ResolveResult:
    parsed_url = urlparse(registry_url)
    if profile.allow_http is False and parsed_url.scheme != "https":
        raise MetadataValidationError("registry_url must use https in strict profile")
    if parsed_url.path.endswith("/.well-known/agent.json"):
        request_url = registry_url
    else:
        base_path = parsed_url.path.rstrip("/")
        if not base_path.endswith("/v1/agents/resolve"):
            base_path += "/v1/agents/resolve"
        request_url = urlunparse(parsed_url._replace(path=base_path, query=urlencode({"agent_id": agent_id})))
    response = await http_client.get(request_url, timeout=timeout_seconds, follow_redirects=False)
    response.raise_for_status()
    payload = response.json()
    if isinstance(payload, dict) and payload.get("registry_type") != "agent_registry":
        metadata_payload = payload.get("metadata", payload)
        metadata = AgentMetadata.model_validate(metadata_payload)
        if metadata.agent_id != agent_id:
            raise MetadataValidationError("registry returned a different agent_id")
        validate_metadata(metadata, profile)
        return ResolveResult(
            metadata=metadata,
            resolved_at=datetime.now(UTC),
            etag=response.headers.get("etag"),
            source_url=request_url,
        )
    document = AgentRegistryDocument.model_validate(payload)
    for entry in document.agents:
        if entry.agent_id != agent_id:
            continue
        validate_metadata(entry.metadata, profile)
        assert_subject_match(agent_id, entry.metadata.domain)
        return ResolveResult(
            metadata=entry.metadata,
            resolved_at=datetime.now(UTC),
            etag=response.headers.get("etag"),
            source_url=request_url,
        )
    raise MetadataValidationError(f"agent not found in registry: {agent_id}")


async def _validate_direct_target(url: str, profile: RuntimeProfile) -> str:
    """在直接发现前阻止本地、私网和保留地址目标。"""

    parsed = urlparse(url)
    hostname = parsed.hostname
    if not hostname:
        raise MetadataValidationError("metadata URL is missing hostname")
    if parsed.username is not None or parsed.password is not None:
        raise MetadataValidationError("metadata URL must not contain userinfo")
    if profile.allow_ip_host:
        return hostname
    if hostname.lower() == "localhost":
        raise MetadataValidationError("localhost metadata target is rejected")
    try:
        literal = ipaddress.ip_address(hostname)
    except ValueError:
        try:
            records = await asyncio.to_thread(socket.getaddrinfo, hostname, parsed.port or 443, type=socket.SOCK_STREAM)
        except OSError as exc:
            raise MetadataValidationError(f"unable to resolve metadata host: {hostname}") from exc
        addresses = {str(record[4][0]) for record in records}
        if not addresses:
            raise MetadataValidationError(f"metadata host has no addresses: {hostname}") from None
        if any(not ipaddress.ip_address(address).is_global for address in addresses):
            raise MetadataValidationError("metadata host resolves to a non-global address") from None
        return sorted(addresses, key=lambda address: (ipaddress.ip_address(address).version, address))[0]
    else:
        if not literal.is_global:
            raise MetadataValidationError("metadata target must use a global address")
        return literal.compressed


def _pin_direct_url(url: str, pinned_ip: str) -> str:
    parsed = urlparse(url)
    hostname = parsed.hostname
    if hostname is None or pinned_ip == hostname:
        return url
    ip = ipaddress.ip_address(pinned_ip)
    host = f"[{ip.compressed}]" if ip.version == 6 else ip.compressed
    if parsed.port is not None:
        host = f"{host}:{parsed.port}"
    return urlunparse(parsed._replace(netloc=host))


def select_verification_key(
    metadata: AgentMetadata,
    *,
    kid: str,
    now: datetime,
) -> AgentKey:
    reference_now = _ensure_aware(now)
    if kid in metadata.revoked_kids:
        raise MetadataValidationError(f"revoked key: {kid}")
    for key in metadata.keys:
        if key.kid != kid:
            continue
        if key.status != "active":
            continue
        not_before = _ensure_aware(key.not_before) if key.not_before else None
        not_after = _ensure_aware(key.not_after) if key.not_after else None
        if not_before and not_before > reference_now:
            raise MetadataValidationError(f"key not active yet: {kid}")
        if not_after and not_after < reference_now:
            raise MetadataValidationError(f"key expired: {kid}")
        return key
    raise MetadataValidationError(f"verification key not found: {kid}")


def _ensure_aware(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=UTC)
    return value.astimezone(UTC)
